src/ray/core_worker/object_recovery_manager.h

```diff
@@ -0,0 +1,151 @@
+// limitations under the License.
+
+#ifndef RAY_CORE_WORKER_OBJECT_RECOVERY_MANAGER_H
+#define RAY_CORE_WORKER_OBJECT_RECOVERY_MANAGER_H
+
+#include "absl/base/thread_annotations.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/common/id.h"
+#include "ray/core_worker/reference_count.h"
+#include "ray/core_worker/store_provider/memory_store/memory_store.h"
+#include "ray/core_worker/task_manager.h"
+#include "ray/raylet/raylet_client.h"
+
+namespace ray {
+
+typedef std::function<std::shared_ptr<PinObjectsInterface>(const std::string &ip_address,
+                                                           int port)>
+    ObjectPinningClientFactoryFn;
+
+typedef std::function<void(const ObjectID &object_id,
+                           const std::vector<rpc::Address> &raylet_locations)>
+    ObjectLookupCallback;
+
+class ObjectRecoveryManager {
+ public:
+  ObjectRecoveryManager(const rpc::Address &rpc_address,
+                        ObjectPinningClientFactoryFn client_factory,
+                        std::shared_ptr<PinObjectsInterface> local_object_pinning_client,
+                        std::function<Status(const ObjectID &object_id,
+                                             const ObjectLookupCallback &callback)>
+                            object_lookup,
+                        std::shared_ptr<TaskResubmissionInterface> task_resubmitter,
+                        std::shared_ptr<ReferenceCounter> reference_counter,
+                        std::shared_ptr<CoreWorkerMemoryStore> in_memory_store,
+                        std::function<void(const ObjectID &object_id, bool pin_object)>
+                            reconstruction_failure_callback,
+                        bool lineage_reconstruction_enabled)
+      : rpc_address_(rpc_address),
+        client_factory_(client_factory),
+        local_object_pinning_client_(local_object_pinning_client),
+        object_lookup_(object_lookup),
+        task_resubmitter_(task_resubmitter),
+        reference_counter_(reference_counter),
+        in_memory_store_(in_memory_store),
+        reconstruction_failure_callback_(reconstruction_failure_callback),
+        lineage_reconstruction_enabled_(lineage_reconstruction_enabled) {}
+
+  /// Recover an object that was stored in plasma. This will only succeed for
+  /// objects that are lost from memory and that this process owns (returns
+  /// Status::Invalid if false).  This method is idempotent for overlapping
+  /// recovery operations on the same object. This class will guarantee that
+  /// each recovery operation ends in either success (by storing a new value
+  /// for the object in the direct memory/plasma store) or failure (by calling
+  /// the given reconstruction failure callback).
+  ///
+  /// Algorithm:
+  /// 1. Check that the object is missing from the direct memory store and that
+  /// we own the object. If either is false, then fail the recovery operation.
```

Comment:
Why not delegate non-owned object recovery to the owner instead of failing? Seems like we have to do that for recursive arg dependencies anyways.

Reply:
I think we should do it, but I thought it'd be better to save that for a future PR, just to keep the changes incremental. This PR only covers reconstruction of objects where the object's lineage is all owned by one process. Reconstruction for borrowed objects is more complicated because we'd have to modify the lineage ref counting protocol to include borrowers' ref counts. I'll update the PR description to account for this.